### DISCUSSION TITLE
Resolved: Hex crate

### DISCUSSION THREAD

**Discussion by Kixunil** [rust-bitcoin/rust-bitcoin#1568] [Refactoring]
There's at least one downstream crate interested in a proper hex crate: https://github.com/stratum-mining/stratum/issues/375

Maybe we should separate it after all? It's not a super-high priority for me now but something to consider. I'm willing to co-maintain it at some point.

**Comment by apoelstra** [comment]
It looks like they'd be okay if we depended on the 3rd party `hex` crate and re-exported it. I'd be willing to do that once we have a crev policy in place.

That also won't be soon, but I think it'll be sooner than us creating our own crate and agreeing on the API and separation of concerns :)

**Reply by Kixunil** [reply]
I'm not sure the comment there meant to say this, they specifically want to remove the `hex` crate.

**Reply by apoelstra** [reply]
Interesting. I understood that, but thought they meant they want to remove their *direct dependency* on `hex`, and would be okay depending on it indirectly through rust-bitcoin, since presumably in that case we'd vetted it. They already depend on other third-party libraries this way anyway.

But actually, now that I'm looking at `cargo tree` on this library ... we **don't** depend on any other third-party libraries. Just `rand` and `serde` which are both optional, and some dev-dependencies. That's actually really cool! For some reason I thought our dep tree was much bigger and less manageable.

**Reply by Kixunil** [reply]
Well, there's stuff like `cc` which we depend on from `secp256k1` build script.

**Comment by dpc** [comment]
In Fedimint we want to [remove hex crate and were hoping to use `bitcoin_hashes` instead](https://github.com/fedimint/fedimint/pull/1349).  Ping @mxz42

**Reply by mxz42** [reply]
Yes, we want to *remove* the `hex` crate and depend on `bitcoin_hashes::hex`. I did a bunch of work and used the `bitcoin_hashes::hex::ToHex` trait for that. Today I saw you removed that trait ~2 weeks ago :sweat_smile: 
So it would be fantastic if we could depend on your optimized hex module. 

2) Another thought was to add the functionality to de/serialize with serde via `#[serde(with = "module")] ` ([serde docs](https://serde.rs/field-attrs.html)). The `hex` crate supports this [with their serde module](https://docs.rs/hex/latest/hex/serde/index.html):

```rust
#[derive(Debug, Serialize, Deserialize)]
pub struct Thing {
    #[serde(with = "hex::serde")]
    pub payload: Vec<u8>,
}
```

Currently, we have a temporary workaround with our [own hex serde module](https://github.com/fedimint/fedimint/blob/5a8000bf9cd8189054c253e89981e4d6a3f2a038/fedimint-api/src/hex/serde.rs) that can be used similar to `hex::serde`:

```rust
#[derive(Debug, Serialize, Deserialize)]
pub struct Thing {
    #[serde(with = "fedimint_api::hex::serde")]
    pub payload: Vec<u8>,
}
```

The plan was to upstream this to `bitcoin_hashes`. Have you thought about such functionality?

I'd also be happy to help if I can.

**Reply by Kixunil** [reply]
Oh, that's unfortunate. I guess we could prioritize it a bit more than I thought. Regarding `serde`, sure, that'd be nice.

**Reply by tcharding** [reply]
@mxz42 and @dpc I've hacked up a new crate, including example usage - for your viewing pleasure :)

https://github.com/rust-bitcoin/rust-bitcoin/pull/1573

**Reply by dpc** [reply]
Thank you!

**Comment by tcharding** [comment]
I'm marking this as resolved because of https://github.com/rust-bitcoin/hex-conservative (FTR there is some chaos still because 1.0.0 only does hex -> bytes. To get bytes -> hex one still has to use pre-1.0 (eg `v0.3`).
